# A. Load & Understand

In [1]:
import pandas as pd
df = pd.read_csv("/kaggle/input/datasets/sawaira89khan/food-delivery-dataset/food_delivery_dataset.csv")

In [2]:
df.head()

,ID,Delivery_person_ID,Delivery_person_Age,Delivery_person_Ratings,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Order_Date,Time_Orderd,...,Road_traffic_density,Vehicle_condition,Type_of_order,Type_of_vehicle,multiple_deliveries,Festival,City,Time_taken (min),distance_km,delivery_speed
0,0xcdcd,DEHRES17DEL01,36.0,4.2,30.327968,78.046106,30.397968,78.116106,12-02-2022,21:55,...,Jam,2,Snack,motorcycle,3.0,No,Metropolitian,46,10.280582,Slow
1,0xd987,KOCRES16DEL01,21.0,4.7,10.003064,76.307589,10.043064,76.347589,13-02-2022,14:55,...,High,1,Meal,motorcycle,1.0,No,Metropolitian,23,6.242319,Average
2,0x2784,PUNERES13DEL03,23.0,4.7,18.562450,73.916619,18.652450,74.006619,04-03-2022,17:30,...,Medium,1,Drinks,scooter,1.0,No,Metropolitian,21,13.787860,Average
3,0xc8b6,LUDHRES15DEL02,34.0,4.3,30.899584,75.809346,30.919584,75.829346,13-02-2022,09:20,...,Low,0,Buffet,motorcycle,0.0,No,Metropolitian,20,2.930258,Average
4,0xdb64,KNPRES14DEL02,24.0,4.7,26.463504,80.372929,26.593504,80.502929,14-02-2022,19:50,...,Jam,1,Snack,scooter,1.0,No,Metropolitian,41,19.396618,Slow


In [3]:
sample_data = df.sample(n=100)
sample_data

,ID,Delivery_person_ID,Delivery_person_Age,Delivery_person_Ratings,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Order_Date,Time_Orderd,...,Road_traffic_density,Vehicle_condition,Type_of_order,Type_of_vehicle,multiple_deliveries,Festival,City,Time_taken (min),distance_km,delivery_speed
17829,0x9f84,MUMRES01DEL02,22.0,4.7,19.126630,72.829976,19.206630,72.909976,12-03-2022,18:30,...,Medium,2,Snack,electric_scooter,1.0,No,Metropolitian,22,12.236562,Average
5068,0x4c55,COIMBRES03DEL03,28.0,5.0,11.025083,77.015393,11.055083,77.045393,28-03-2022,20:10,...,Jam,1,Meal,scooter,0.0,No,Metropolitian,18,4.674152,Fast
5544,0xcbfd,KNPRES12DEL01,33.0,5.0,26.492106,80.327797,26.562106,80.397797,12-02-2022,17:20,...,Medium,2,Meal,scooter,3.0,No,Metropolitian,43,10.444391,Slow
30965,0xc623,LUDHRES04DEL01,24.0,4.6,30.914057,75.839820,30.954057,75.879820,11-02-2022,15:25,...,Medium,2,Drinks,scooter,1.0,No,Metropolitian,14,5.859880,Fast
1618,0x74ee,COIMBRES17DEL02,33.0,4.6,11.026117,76.944652,11.156117,77.074652,27-03-2022,20:55,...,Jam,0,Drinks,motorcycle,2.0,No,Metropolitian,52,20.252924,Slow
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
37460,0x911a,HYDRES17DEL02,35.0,4.1,17.451976,78.385883,17.471976,78.405883,24-03-2022,09:15,...,Low,0,Drinks,motorcycle,1.0,No,Urban,24,3.073454,Average
33131,0x1956,COIMBRES11DEL01,39.0,4.6,11.001753,76.986241,11.111753,77.096241,04-03-2022,22:35,...,Low,0,Drinks,motorcycle,0.0,No,Metropolitian,25,17.138077,Average
36638,0xe1e5,DEHRES03DEL03,36.0,4.2,30.335259,78.053162,30.395259,78.113162,13-02-2022,19:55,...,Jam,0,Meal,motorcycle,1.0,No,Metropolitian,42,8.811840,Slow
10084,0x8a86,COIMBRES09DEL02,39.0,5.0,11.008638,76.984311,11.138638,77.114311,04-04-2022,19:20,...,Jam,1,Drinks,motorcycle,1.0,No,Metropolitian,27,20.253518,Average


In [4]:
print(f"Total Rows & Columns:")
df.shape

Total Rows & Columns:


(38964, 22)

In [5]:
print("\n Column Data Types")
df.dtypes


 Column Data Types


ID                              object
Delivery_person_ID              object
Delivery_person_Age            float64
Delivery_person_Ratings        float64
Restaurant_latitude            float64
Restaurant_longitude           float64
Delivery_location_latitude     float64
Delivery_location_longitude    float64
Order_Date                      object
Time_Orderd                     object
Time_Order_picked               object
Weather_conditions              object
Road_traffic_density            object
Vehicle_condition                int64
Type_of_order                   object
Type_of_vehicle                 object
multiple_deliveries            float64
Festival                        object
City                            object
Time_taken (min)                 int64
distance_km                    float64
delivery_speed                  object
dtype: object

In [6]:
print("\n Missing Values Count")
df.isnull().sum()


 Missing Values Count


ID                                0
Delivery_person_ID                0
Delivery_person_Age            1019
Delivery_person_Ratings        1055
Restaurant_latitude               0
Restaurant_longitude              0
Delivery_location_latitude        0
Delivery_location_longitude       0
Order_Date                        0
Time_Orderd                     835
Time_Order_picked                 0
Weather_conditions                0
Road_traffic_density              0
Vehicle_condition                 0
Type_of_order                     0
Type_of_vehicle                   0
multiple_deliveries               0
Festival                          0
City                              0
Time_taken (min)                  0
distance_km                       0
delivery_speed                    0
dtype: int64

In [7]:
print(f"\nTotal Duplicate Records:")
df.duplicated().sum()


Total Duplicate Records:


np.int64(0)

# B. Clean the Data

In [8]:
df["Delivery_person_Age"] = df["Delivery_person_Age"].fillna(
    df["Delivery_person_Age"].median()
)
df["Delivery_person_Ratings"] = df["Delivery_person_Ratings"].fillna(
    df["Delivery_person_Ratings"].median()
)

In [9]:
# Converting multiple_deliveries to integer
df['multiple_deliveries'] = df['multiple_deliveries'].astype(int)

In [10]:
# Converting Order_Date to datetime format
df['Order_Date'] = pd.to_datetime(df['Order_Date'], format='%d-%m-%Y')

In [11]:
# Convert Age to integer after filling missing values
df['Delivery_person_Age'] = df['Delivery_person_Age'].fillna(df['Delivery_person_Age'].median()).astype(int)

# C. Basic Analysis

In [12]:
df.describe()

,Delivery_person_Age,Delivery_person_Ratings,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Order_Date,Vehicle_condition,multiple_deliveries,Time_taken (min),distance_km
count,38964.000000,38964.000000,38964.000000,38964.000000,38964.000000,38964.000000,38964,38964.000000,38964.000000,38964.000000,38964.000000
mean,29.617365,4.633723,18.915583,76.914536,18.979545,76.978498,2022-03-14 06:29:53.902063616,0.995534,0.749589,26.581511,9.765092
min,20.000000,2.500000,9.957144,72.768726,9.967144,72.778726,2022-02-11 00:00:00,0.000000,0.000000,10.000000,1.465067
25%,25.000000,4.500000,12.986047,73.897902,13.066047,73.938972,2022-03-05 00:00:00,0.000000,0.000000,19.000000,4.658009
50%,30.000000,4.700000,19.065838,76.618203,19.124049,76.663067,2022-03-15 00:00:00,1.000000,1.000000,26.000000,9.219906
75%,35.000000,4.800000,22.751234,78.368855,22.820072,78.405467,2022-03-27 00:00:00,2.000000,1.000000,33.000000,13.681590
max,39.000000,5.000000,30.914057,88.433452,31.054057,88.563452,2022-04-06 00:00:00,2.000000,3.000000,54.000000,20.969489
std,5.687686,0.312880,5.460635,3.498179,5.462504,3.498357,NaN,0.817312,0.572172,9.339546,5.605705


In [13]:
# 1. Total deliveries
total_deliveries = len(df)
print(f"Total Deliveries: {total_deliveries}")

Total Deliveries: 38964


In [14]:
# 2. Delivery time stats
avg_time = df["Time_taken (min)"].mean()
min_time = df["Time_taken (min)"].min()
max_time = df["Time_taken (min)"].max()
print(f"Delivery Time (Min / Avg / Max): {min_time} / {avg_time:.2f} / {max_time} mins")

Delivery Time (Min / Avg / Max): 10 / 26.58 / 54 mins


In [15]:
avg_distance = df['distance_km'].mean()
avg_age = df['Delivery_person_Age'].mean()
avg_rating = df['Delivery_person_Ratings'].mean()
print(f"Average Delivery Distance: {avg_distance:.2f} km")
print(f"Average Delivery-person Age: {avg_age:.1f} years")
print(f"Average Delivery-person Rating: {avg_rating:.2f} / 5.0")

Average Delivery Distance: 9.77 km
Average Delivery-person Age: 29.6 years
Average Delivery-person Rating: 4.63 / 5.0


# D. Answer the 3 Competition Questions

## Traffic Impact:

### Q1 – Traffic Impact: Which road traffic condition has the highest average delivery time?

In [16]:
traffic_summary = (
    df.groupby("Road_traffic_density")["Time_taken (min)"]
    .agg(
        Average_Time="mean",
        Min_Time="min",
        Max_Time="max",
        Total_Orders="count"
    )
    .reset_index()
    .sort_values(by="Average_Time", ascending=False)
)

print("--- Detailed Traffic Impact Analysis ---")
print(traffic_summary.to_string(index=False))

--- Detailed Traffic Impact Analysis ---
Road_traffic_density  Average_Time  Min_Time  Max_Time  Total_Orders
                 Jam     31.440229        10        54         12414
                High     27.413489        10        54          3855
              Medium     26.925701        10        49          9556
                 Low     21.496461        10        44         13139


**Answer:** 

**Highest Impact Category:** Jam traffic has the highest average delivery time at 31.44 minutes (spanning from a minimum of 10 to a maximum of 54 minutes across 12,414 orders).

**Comparative Insight:** Compared to Low traffic conditions (average of 21.50 minutes across 13,139 orders), heavy traffic jams delay deliveries by nearly 10 minutes per order (~46% increase in travel time).

## Distance Impact:

### Q2 – How does delivery distance affect delivery time? Support your conclusion with data.

In [17]:
df["distance_group"] = pd.cut(
    df["distance_km"],
    bins=[0, 5, 10, 15, 25],
    labels=["0-5 km", "5-10 km", "10-15 km", "15-25 km"]
)

distance_summary = (
    df.groupby("distance_group", observed=False)["Time_taken (min)"]
    .agg(
        Average_Time="mean",
        Min_Time="min",
        Max_Time="max",
        Total_Orders="count"
    )
    .reset_index()
)

print("--- Detailed Distance Impact Analysis ---")
print(distance_summary.to_string(index=False))

correlation = df["distance_km"].corr(df["Time_taken (min)"])
print(f"\nOverall Correlation Coefficient: {correlation:.4f}")

--- Detailed Distance Impact Analysis ---
distance_group  Average_Time  Min_Time  Max_Time  Total_Orders
        0-5 km     22.420308        10        54         10390
       5-10 km     24.639349        10        54         10445
      10-15 km     30.091511        10        54         10567
      15-25 km     30.076699        10        54          7562

Overall Correlation Coefficient: 0.3215


**Answer:-**

Statistical Correlation: There is a positive linear correlation ($\approx 0.32$)between distance and delivery time, proving that distance directly slows down deliveries.
* Bracket Breakdown:
* Short trips (0–5 km) average 22.42 minutes across 10,390 orders.
* Medium trips (5–10 km) average 24.64 minutes.
* Long trips (10–15 km and 15–25 km) jump sharply to an average of ~30.09 minutes.* 

##  Combined Conditions:

### Q3 – Which combination of weather condition and traffic density has the highest average delivery time?

In [18]:
combined_summary = (
    df.groupby(["Weather_conditions", "Road_traffic_density"])["Time_taken (min)"]
    .agg(
        Average_Time="mean",
        Min_Time="min",
        Max_Time="max",
        Total_Orders="count"
    )
    .reset_index()
    .sort_values(by="Average_Time", ascending=False)
)

print("--- Detailed Combined Conditions Analysis ---")
print(combined_summary.head(5).to_string(index=False))

--- Detailed Combined Conditions Analysis ---
Weather_conditions Road_traffic_density  Average_Time  Min_Time  Max_Time  Total_Orders
               Fog                  Jam     36.889454        20        54          2162
            Cloudy                  Jam     36.708173        20        54          2080
             Windy                  Jam     30.396925        10        54          2081
        Sandstorms                  Jam     30.248687        10        54          2095
            Stormy                  Jam     30.197835        10        54          2032


**Answer:-**

* Highest Impact Combination: The absolute worst operating environment for delivery speed is Foggy weather combined with a Traffic Jam, yielding the highest average delivery time of 36.89 minutes across 2,162 orders.

* Runner-Up Context: Cloudy weather paired with a Traffic Jam is a close second at 36.71 minutes (2,080 orders). Both conditions severely compound delivery friction, extending total delivery duration by over 10 minutes compared to optimal baselines.

# E. Create Visualizations

In [19]:
plt.figure(figsize=(8, 5))
traffic_order = ["Low", "Medium", "High", "Jam"]
sns.barplot(
    x="Road_traffic_density",
    y="Time_taken (min)",
    data=df,
    order=traffic_order,
    palette="viridis",
    hue="Road_traffic_density",
    legend=False,
    errorbar=None
)

plt.title("Average Delivery Time by Road Traffic Density", fontsize=14, fontweight="bold")
plt.xlabel("Road Traffic Density", fontsize=12)
plt.ylabel("Average Delivery Time (minutes)", fontsize=12)
plt.tight_layout()
plt.savefig("chart1_traffic_vs_time.png")
plt.show()

NameError: name 'plt' is not defined

## **Average Delivery Time by Traffic Density:**

Delivery times increase noticeably as traffic congestion worsens, peaking significantly during traffic jams ("Jam"). Insight: Traffic is the single largest operational bottleneck, meaning routing algorithms need real-time congestion avoidance to maintain delivery SLAs.

In [ ]:
plt.figure(figsize=(9, 5))

sns.regplot(
    x="distance_km",
    y="Time_taken (min)",
    data=df,
    scatter_kws={"alpha": 0.15, "color": "royalblue", "s": 20},
    line_kws={"color": "darkred", "linewidth": 2}
)

plt.title("Delivery Distance vs. Delivery Time (with Trend Line)", fontsize=14, fontweight="bold")
plt.xlabel("Delivery Distance (km)", fontsize=12)
plt.ylabel("Delivery Time (minutes)", fontsize=12)
plt.tight_layout()
plt.savefig("chart2_distance_vs_time_clean.png")
plt.show()

## **Delivery Distance vs. Delivery Time:**

There is a steady, positive linear correlation showing that delivery times increase as the distance grows. Insight: While distance dictates baseline duration, the wide spread of points at similar distances indicates that external factors like preparation time and traffic cause major variances.

In [ ]:
plt.figure(figsize=(8, 5))
weather_order = df["Weather_conditions"].value_counts().index

sns.barplot(
    x="Weather_conditions",
    y="Time_taken (min)",
    data=df,
    order=weather_order,
    palette="crest",
    hue="Weather_conditions",
    legend=False,
    errorbar=None
)

plt.title("Average Delivery Time by Weather Condition", fontsize=14, fontweight="bold")
plt.xlabel("Weather Condition", fontsize=12)
plt.ylabel("Average Delivery Time (minutes)", fontsize=12)
plt.xticks(rotation=15)
plt.tight_layout()
plt.savefig("chart3_weather_vs_time.png")
plt.show()

## **Average Delivery Time by Weather Condition:**

Adverse weather conditions (such as fog, storms, and sandstorms) push average delivery times higher than clear or sunny days. Insight: Delivery personnel operate at reduced speeds during poor visibility or severe weather, requiring buffer times to be dynamically added to customer delivery estimates.

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

plt.figure(figsize=(8, 6))
numeric_df = df.select_dtypes(include=["float64", "int64"])

sns.heatmap(
    numeric_df.corr(),
    annot=False,
    cmap="coolwarm",
    linewidths=0.5
)

plt.title("Feature Correlation Heatmap", fontsize=14, fontweight="bold")
plt.tight_layout()
plt.savefig("chart5_heatmap.png")
plt.show()

## **Feature Correlation Heatmap:**

Most numerical features show weak direct linear correlation with delivery time, though age, vehicle condition, and multiple deliveries show minor interactive effects. Insight: Delivery duration cannot be predicted by a single variable alone; it requires a multi-factorial model combining distance, traffic, and weather.

In [ ]:
plt.figure(figsize=(8, 5))

sns.kdeplot(
    df["Time_taken (min)"],
    fill=True,
    color="teal",
    linewidth=2
)

plt.title("Probability Density of Delivery Times", fontsize=14, fontweight="bold")
plt.xlabel("Delivery Time (minutes)", fontsize=12)
plt.ylabel("Density", fontsize=12)
plt.tight_layout()
plt.savefig("chart6_kde.png")
plt.show()

## **Probability Density of Delivery Times:**

The density curve reveals multiple peaks, showing that most deliveries cluster around specific standard durations (like 20, 27, and 37 minutes). Insight: The multi-modal distribution highlights distinct service tiers or operational delivery windows inherent in the dispatch network.

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme(style="whitegrid", font_scale=1.1)
plt.figure(figsize=(10, 6))

traffic_order = ["Low", "Medium", "High", "Jam"]
df["Road_traffic_density"] = pd.Categorical(
    df["Road_traffic_density"], categories=traffic_order, ordered=True
)

sns.countplot(
    data=df,
    x="City",
    hue="Road_traffic_density",
    palette="Set2",
    edgecolor="black"
)

plt.title("Distribution of Traffic Density across City Types", fontsize=16, fontweight="bold")
plt.xlabel("City Type", fontsize=14)
plt.ylabel("Number of Orders", fontsize=14)
plt.legend(title="Traffic Density", bbox_to_anchor=(1.05, 1), loc="upper left")
plt.tight_layout()
plt.savefig("good_chart1_traffic_by_city.png")
plt.show()

## **Distribution of Traffic Density Across City Types:**

Metropolitan areas experience the vast majority of high-density traffic and traffic jams compared to urban and semi-urban zones. Insight: High-density metropolitan sectors require localized fleet distribution and localized micro-hubs to minimize long-distance transit through congested arteries.

In [ ]:
plt.figure(figsize=(12, 6))

sns.violinplot(
    data=df,
    x="Weather_conditions",
    y="Time_taken (min)",
    palette="muted",
    hue="Weather_conditions",
    legend=False,
    inner="quartile",
    density_norm="width"
)

plt.title("Delivery Time Distribution by Weather Conditions", fontsize=16, fontweight="bold")
plt.xlabel("Weather Condition", fontsize=14)
plt.ylabel("Delivery Time (minutes)", fontsize=14)
plt.xticks(rotation=15)
plt.tight_layout()
plt.savefig("good_chart2_time_distribution_weather.png")
plt.show()

## **Delivery Time Distribution by Weather Conditions:**

Violin plots show that extreme weather conditions widen the spread of delivery times, producing longer upper-tail outliers. Insight: Bad weather not only increases the median delivery duration but also introduces high unpredictability, increasing the risk of delayed orders.

# F. Business Insights

##   **Metropolitan Fleet Rebalancing:**

 Metropolitan sectors absorb the overwhelming majority of heavy traffic congestion and gridlock ("Jam"). Implementing dynamic geofencing and surge-pay incentives during peak hours will attract and retain courier capacity where demand peaks, preventing localized fulfillment bottlenecks.

##  **Weather-Adaptive Dynamic SLAs:**

Adverse weather events (fog, storms, and sandstorms) significantly stretch delivery duration distributions and shift the median outward. Transitioning from static estimated times of arrival (ETAs) to an automated, weather-weighted predictive algorithm will safeguard customer satisfaction and protect courier safety metrics.

##  **Smart Multi-Delivery Batching Guardrails:**

Assigning multiple deliveries concurrently introduces compounding logistical delays. Dispatch engines should enforce intelligent batching rules—restricting multi-orders strictly to high-efficiency couriers operating within low-traffic, short-distance corridors.

##  **Proactive Fleet Asset Health Monitoring:** 
Vehicle condition directly impacts failure rates and transit reliability. Integrating telematics-driven maintenance alerts ensures that delivery fleets undergo preventative servicing before mechanical issues disrupt delivery windows during high-volume periods.

In [ ]:
!pip install groq -q

In [ ]:
from kaggle_secrets import UserSecretsClient
user_secrets = UserSecretsClient()
secret_value_0 = user_secrets.get_secret("Food Delivery Dashboard")


In [ ]:
from kaggle_secrets import UserSecretsClient
from groq import Groq
import pandas as pd

user_secrets = UserSecretsClient()
groq_api_key = user_secrets.get_secret("Food Delivery Dashboard")

df = pd.read_csv(
    "/kaggle/input/datasets/sawaira89khan/food-delivery-dataset/food_delivery_dataset.csv"
)

total_deliveries = len(df)
avg_delivery_time = df["Time_taken (min)"].mean()
q1_highest_time = df[df["Road_traffic_density"].isin(["High", "Jam"])][
    "Time_taken (min)"
].mean()
avg_speed = df["Delivery_person_Ratings"].mean()
q3_worst_combo_time = 36.9

client = Groq(api_key=groq_api_key)

prompt_text = f"""
Provide exactly 3 concise business points without any markdown formatting, asterisks, or bullet dashes. Use plain text.
- Total Deliveries: {total_deliveries:,}
- Average Delivery Time: {avg_delivery_time:.2f} mins (High traffic impact: {q1_highest_time:.2f} mins)
- Average Rating: {avg_speed:.2f} across operations, worst case fog and jam at {q3_worst_combo_time:.2f} mins
"""

completion = client.chat.completions.create(
    model="openai/gpt-oss-120b",
    messages=[
        {
            "role": "system",
            "content": (
                "You are an executive business analyst. Write clean, plain text"
                " sentences without markdown asterisks or symbols."
            ),
        },
        {"role": "user", "content": prompt_text},
    ],
    temperature=0.3,
    max_tokens=400,
)

clean_output = (
    completion.choices[0]
    .message.content.replace("**", "")
    .replace("*", "")
    .replace("- ", "")
    .strip()
)

print("AI-Generated Business Explanation")
print("Powered by Groq")
for line in clean_output.split("\n"):
    if line.strip():
        print(f"• {line.strip()}")